XMonad/Hooks/StatusBar/PP.hs

```diff
@@ -200,17 +210,105 @@ pprWindowSet :: WorkspaceSort -> [Window] -> PP -> WindowSet -> String
-                | isJust (S.stack w)                                      = ppHidden
-                | otherwise                                               = ppHiddenNoWindows
+        printer | Just (_, f) <- find (\pf ->  fst pf urgents s w) preds = f
+                | otherwise                                              = ppHiddenNoWindows
+
+-- $predicates
+-- Using 'WorkspacePredicate' and 'WorkspaceFormatter' with
+-- 'ppPrinters' can allow users to define new workspace types beyond
+-- current, urgent, and the likes that are defined in 'PP' fields. The
+-- 'WorkspaceFormatter' corresponding to the first
+-- 'WorkspacePredicate' that returns 'True' in 'ppPrinters' is used to
+-- format the 'WorkspaceId'. If nothing matches, the default workspace
+-- formatters are used.
+
+type WorkspaceFormatter = PP -> WorkspaceId -> String
+type WorkspacePredicate = [Window] -- ^ Urgent windows
+        -> WindowSet
+        -> WindowSpace
+        -> Bool
+
+-- | Predicate for urgent workspaces.
+isUrgent :: WorkspacePredicate
+isUrgent urgents s w = any (\x -> (== Just (S.tag w)) (S.findTag x s)) urgents
+
+-- | Predicate for the current workspace. Caution: assumes default
+-- precedence is respected.
+isCurrent' :: WorkspacePredicate
+isCurrent' _ s w = S.tag w == S.currentTag s
+
+-- | Predicate for the current workspace.
+isCurrent :: WorkspacePredicate
+isCurrent = notWP isUrgent `andWP` isCurrent'
+
+-- | Predicate for visible workspaces. Caution: assumes default
+-- precedence is respected.
+isVisible' :: WorkspacePredicate
+isVisible' _ s w = isVisibleNoWindows' [] s w && isJust (S.stack w)
+
+-- | Predicate for visible workspaces.
+isVisible :: WorkspacePredicate
+isVisible = notWP isUrgent `andWP` notWP isCurrent' `andWP` isVisible'
+
+-- | Predicate for visible workspaces that have no windows. Caution:
+-- assumes default precedence is respected.
+isVisibleNoWindows' :: WorkspacePredicate
+isVisibleNoWindows' _ s w = S.tag w `elem` visibles
+  where visibles = map (S.tag . S.workspace) (S.visible s)
+
+-- | Predicate for visible workspaces that have no windows.
+isVisibleNoWindows :: WorkspacePredicate
+isVisibleNoWindows =
+    notWP isUrgent
+        `andWP` notWP isCurrent'
+        `andWP` notWP isVisible'
+        `andWP` isVisibleNoWindows'
+
+-- | Predicate for hidden workspaces. Caution: assumes default
+-- precedence is respected.
+isHidden' :: WorkspacePredicate
+isHidden' _ _ = isJust . S.stack
+
+-- | Predicate for hidden workspaces.
+isHidden :: WorkspacePredicate
+isHidden =
+    notWP isUrgent
+        `andWP` notWP isCurrent'
+        `andWP` notWP isVisible'
+        `andWP` notWP isVisibleNoWindows'
+        `andWP` isHidden'
+
+-- | Negates a 'WorkspacePredicate'.
+notWP :: WorkspacePredicate -> WorkspacePredicate
+notWP p = not .:. p
+  where  (.:.) = (.) . (.:)
+
+-- | '&&' for 'WorkspacePredicate's.
+andWP :: WorkspacePredicate -> WorkspacePredicate  -> WorkspacePredicate
+andWP = liftM2x3 (&&)
```

Comment:
This is kind of fun because a stacking of `liftA2`s seems completely over the top for this, and yet it seems much nicer to read than

``` haskell
wPred `andWP` wPred' = \wins winSet winSpace -> wPred  wins winSet winSpace 
                                             && wPred' wins winSet winSpace
```

Although I question which one is clearer a bit :)

Reply:
Either way is good for me, I was just practicing some Haskell wizardry

Reply:
Eh, probably not that important; just go with your gut I suppose :)